Migrate good practice for container images into Containers section (#29345)
kubernetes/website: content/en/docs/concepts/containers/images.md

{{< glossary_tooltip text="kubelet" term_id="kubelet" >}} to skip pulling an
image if it already exists.

### Image pull policy

The `imagePullPolicy` for a container and the tag of the image affect when the
[kubelet](/docs/reference/command-line-tools-reference/kubelet/) attempts to pull the specified image.

Here's a list of the values you can set for `imagePullPolicy` and the effect

> that it has:

Reviewer: and the effects these values have:

> Here's a list of the values you can set for `imagePullPolicy` and the effect
> that it has:

Author: tip: use a suggestion when proposing a direct replacement, eg
```suggestion
Here's a list of the values you can set for `imagePullPolicy` and the effects
these values have:
```
(feedback on https://github.com/kubernetes/website/pull/29345/files#r688073241)

`IfNotPresent`
: the image is pulled only if it is not already present locally.

`Always`
: every time the kubelet launches a container, the kubelet queries the container

>   image registry to resolve the name to an image digest. If the kubelet has a

Reviewer: is the user supposed to know what a digest is?

Author: We could add _image digest_ to the glossary. Personally, I'd save that for a follow-up PR.

  container image with that exact digest cached locally, the kubelet uses its cached

>   image; otherwise, the kubelet downloads (pulls) the image with the resolved digest,

Reviewer: I'd just say pulls here. If you want to have "downloads" in there, maybe add it to the intro sentence for this section "attempts to download (pull) the specified image."

  and uses that image to launch the container.

`Never`
: the kubelet does not try fetching the image. If the image is somehow already present

>   locally, the kubelet is able to start the container; otherwise, startup fails.

Reviewer: the kubelet can start the container

>   See [pre-pulled images](#pre-pulled-images) for more on this.

Reviewer: for more information/details

The caching semantics of the underlying image provider make even
`imagePullPolicy: Always` efficient, as long as the registry is reliably accessible.
Your container runtime can notice that the image layers already exist on the node
so that they don't need to be downloaded again.

{{< note >}}
You should avoid using the `:latest` tag when deploying containers in production as
it is harder to track which version of the image is running and more difficult to
roll back properly.

Instead, specify a meaningful tag such as `v1.42.0`.
{{< /note >}}

To make sure the Pod always uses the same version of a container image, you can specify

> the image's [digest](https://docs.docker.com/engine/reference/commandline/pull/#pull-an-image-by-digest-immutable-identifier);

Reviewer: This link should be on line 65 ideally as that's the first time digest is used.

replace `<image-name>:<tag>` with `<image-name>@<digest>`
(for example, `image@sha256:45b23dee08af5e43a7fea6c4cf9c25ccf269ee113168c19722f87876677c5cb2`).

The digest uniquely identifies a specific version of the image, so it is never updated by Kubernetes
unless you change the digest value. If the image registry were to change the code that the

> tag on that image represents, you might end up with a mix of Pods running the old and new

Reviewer: This sentence switches to talking about tags a bit abruptly. Perhaps an introductory bit like "When using image tags, if the image registry..."

> code. Specifying an image fixes the code that you run so that a change at the registry cannot

Reviewer: Specifying an image digest fixes..

Reviewer: so that a change in the registry (or by the registry)

lead to that mix of versions.

#### Default image pull policy {#imagepullpolicy-defaulting}

When you (or a controller) submit a new Pod to the API server, your cluster sets the
`imagePullPolicy` field when specific conditions are met:

- if you omit the `imagePullPolicy` field, and the tag for the container image is
  `:latest`, `imagePullPolicy` is automatically set to `Always`.

> - if you omit the `imagePullPolicy` field, and you don't specify the tag for the

Reviewer: missing period

Reviewer 2: For lists, so long as the punctuation is consistent across the list the grammar is accurate, so Tim is doing just fine here.

the pull policy of any object after its initial creation.
{{< /note >}}

#### Required image pull

If you would like to always force a pull, you can do one of the following:

> - set the `imagePullPolicy` of the container to `Always`.

Reviewer: Is there a reason the first letter here is not capitalized in all the list items? Recommend capitalizing.

Author: We usually write this kind of list all lower case, but it makes sense here because each item is a sentence. Good spot.
